docs/architecture/adr-003-values-propagation.md

```diff
@@ -0,0 +1,351 @@
+# ADR 003: Propagation of Proposed Values
+
+## Changelog
+
+* 2025-03-18: Context and description of the problem
+* 2025-03-21: Current design description
+
+## Context
+
+Malachite implements a consensus algorithm, [Tendermint][consensus-spec], 
+which allow processes to agree on a single decided value for each 
+consensus instance or *height*. 
+
+Possible decision values are provided by the software that uses the consensus, 
+which we refer to generically as the *application*. There are 
+no assumptions about what a **value** represents—its semantics are defined 
+entirely by the application. For example, in blockchain applications, input 
+values are blocks proposed to be appended to the blockchain.
+
+Similarly, the consensus algorithm makes no assumptions about the **size** 
+of the proposed values; they may be of arbitrary byte size. However, 
+the application is expected to define a maximum size for proposed values 
+and to configure the consensus parameters accordingly—most notably, 
+the durations of timeouts.
+
+When value size becomes a relevant factor, it is important to recognize 
+that the consensus process comprises two distinct stages:
+
+- **Value Propagation**: Proposed values must be transmitted to all consensus participants.
+- **Value Decision**: One of the successfully propagated values is selected and decided.
+
+The cost of **Value Propagation**, in terms of latency and bandwidth, 
+clearly depends on the size of the proposed values. In contrast, the 
+**Value Decision** stage should be independent of value size and 
+incur a roughly constant cost.
+
+In Tendermint, **Value Propagation** is performed via the `PROPOSAL` 
+message, which is broadcast by the designated proposer of the round 
+and includes the proposed value `v`.
+
+The **Value Decision** phase involves `PREVOTE` and `PRECOMMIT` 
+messages—collectively referred to as *votes*. Each vote includes 
+either an identifier `id(v)` of the proposed value `v` or the 
+special value `nil`. The function `id(v)` provides a compact 
+representation of `v`, typically implemented as a fixed-size hash.
+
+From this, we can see that **Value Propagation** is more challenging, 
+as it involves disseminating potentially large amounts of data through 
+the network. In contrast, the **Value Decision** phase requires only 
+the transmission of vote messages, which are relatively small and of 
+constant size. As a result, Malachite’s low-level API provides greater 
+flexibility in the value propagation stage to enable more optimized 
+implementations. 
+
+In this document, we focus on the core parts of the consensus 
+implementation responsible for **Value Propagation**.
+
+## Current design
```

Comment:
We should first define the blocks used in the following:
- Consensus core
- Network
- Application
- Consensus engine (= Core + network?)